### PR TITLE
Cast address type in protobuf

### LINKED ISSUE #482: Cast address type in protobuf
Older protobuf definitions declare addresses as `[]byte` ([example](https://github.com/iov-one/weave/blob/master/x/cash/codec.proto#L19-L20)). Use gogoproto extension to cast those values to `weave.Address` for generated Go code ([example](https://github.com/iov-one/weave/blob/master/x/distribution/codec.proto#L15)).

This change will improve readability and remove the need for [explicit casting](https://github.com/iov-one/weave/blob/master/x/cash/msg.go#L35-L40) of `[]byte` to `weave.Address`.

---

Other extensions might need additional type casting. For example [`x/escrow`](https://github.com/iov-one/weave/blob/master/x/escrow/codec.proto#L17-L20)

### PR DESCRIPTION
**Pull request by husio**
`x/cash`, `x/escrow`: update protobuf message types
    
Cast all message attributes to their respective Go types if needed. For example cast weave address from `[]bytes` to `weave.Address`.
    
resolve #482

### CODE REVIEW

File: Makefile [hunk 1]
```diff
@@ -58,7 +58,7 @@ ifndef $(shell command -v prototool help > /dev/null)
 endif
 	prototool lint
 
-protoc: protodocs
+protoc: #protodocs
```
Comment (alpe) [inline Makefile]: req: please make sure that you not reverting the protodocs change. It is important for our clients.

File: conditions.go [hunk 2]
```diff
@@ -70,6 +70,9 @@ func (c Condition) String() string {
 
 // Validate returns an error if the Condition is not the proper format
 func (c Condition) Validate() error {
+	if len(c) == 0 {
```
Comment (alpe) [inline conditions.go]: 👍

File: conditions.go [hunk 3]
```diff
@@ -201,6 +204,9 @@ func (a Address) String() string {
 
 // Validate returns an error if the address is not the valid size
 func (a Address) Validate() error {
+	if len(a) == 0 {
```
Comment (alpe) [inline conditions.go]: 👍

File: x/cash/handler_test.go [hunk 4]
```diff
@@ -25,68 +22,87 @@ func TestSend(t *testing.T) {
...
-			_, err := h.Check(nil, kv, tx)
-			assert.True(t, tc.expectCheck(err), "%+v", err)
-			_, err = h.Deliver(nil, kv, tx)
-			assert.True(t, tc.expectDeliver(err), "%+v", err)
+			if _, err := h.Check(nil, kv, tx); !tc.wantCheckErr.Is(err) {
+				t.Fatalf("unexpected check error: %+v", err)
+			}
+			if _, err := h.Deliver(nil, kv, tx); !tc.wantDeliverErr.Is(err) {
+				t.Fatalf("unexpected deliver error: %+v", err)
+			}
```
Comment (alpe) [inline x/cash/handler_test.go]: 👍

File: x/cash/msg_test.go [hunk 5]
```diff
@@ -1,138 +1,149 @@
...
+				Ref:    []byte("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaB"),
+			},
+			wantErr: errors.ErrInvalidState,
+		},
+		"memo too long": {
+			msg: &SendMsg{
+				Amount: coin.NewCoinp(10, 0, "FOO"),
+				Dest:   addr1,
+				Src:    addr2,
+				Ref:    []byte("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaB"),
```
Comment (alpe) [inline x/cash/msg_test.go]: req: this not the memo

File: x/cash/msg_test.go [hunk 6]
```diff
@@ -1,138 +1,149 @@
...
+				Src:    addr2,
+			},
+			wantErr: errors.ErrEmpty,
+		},
+		"reference too long": {
+			msg: &SendMsg{
+				Amount: coin.NewCoinp(10, 0, "FOO"),
+				Dest:   addr1,
+				Src:    addr2,
+				Ref:    []byte("aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaB"),
```
Comment (alpe) [inline x/cash/msg_test.go]: q: how many chars are this? I find this harder to read than calling some function that creates the string. for example bigString(128)
q: is there any reason for `B` at the end?

File: x/cash/msg_test.go [hunk 7]
```diff
@@ -1,138 +1,149 @@
...
-	err = minus.Validate()
-	assert.Error(t, err)
-	assert.True(t, errors.ErrInvalidAmount.Is(err))
+}
 
-	bad := coin.NewCoin(3, 0, "fab9")
-	ugly := &FeeInfo{
-		Payer: addr,
-		Fees:  &bad,
+func randomAddr(t testing.TB) weave.Address {
```
Comment (alpe) [inline x/cash/msg_test.go]: pp: you can replace this by `weavetest.NewCondition().Address()` or `weavetest.NewKey().PublicKey().Address()` or just move this into weavetest package

File: Makefile (hunk 1 above)
Comment (husio) [reply Makefile]: Sorry for that, I always forget to uncomment it :/

File: x/cash/msg_test.go (hunk 5 above)
Comment (husio) [reply x/cash/msg_test.go]: Great catch! :mag:

File: x/cash/msg_test.go (hunk 6 above)
Comment (husio) [reply x/cash/msg_test.go]: I have updated to use `strings.Repeat` to make the characters count easier.

File: x/cash/msg_test.go (hunk 7 above)
Comment (husio) [reply x/cash/msg_test.go]: https://github.com/iov-one/weave/pull/526/commits/4586c56268bbbb8130b2767b0f1796214ff038ca

**Pull request merged**